awx/api/views/__init__.py

```diff
@@ -710,16 +712,81 @@ def get(self, request):
         return Response(data)
 
 
+def immutablesharedfields(cls):
+    '''
+    Class decorator to prevent modifying shared resources when AWX_DIRECT_SHARED_RESOURCE_MANAGEMENT_ENABLED setting is set to False.
+
+    Works by overriding these view methods:
+    - create
+    - delete
+    - perform_update
+    create and delete are overridden to raise a PermissionDenied exception.
+    perform_update is overridden to check if any shared fields are being modified,
+    and raise a PermissionDenied exception if so.
+    '''
+    # create instead of perform_create because some of our views
+    # override create instead of perform_create
+    if hasattr(cls, 'create'):
+        cls.original_create = cls.create
+
+        @functools.wraps(cls.create)
+        def create_wrapper(*args, **kwargs):
+            if settings.AWX_DIRECT_SHARED_RESOURCE_MANAGEMENT_ENABLED:
+                return cls.original_create(*args, **kwargs)
+            raise PermissionDenied({'detail': _('Creation of this resource is not allowed. Create this resource via the platform ingress.')})
+
+        cls.create = create_wrapper
+
+    if hasattr(cls, 'delete'):
+        cls.original_delete = cls.delete
+
+        @functools.wraps(cls.delete)
+        def delete_wrapper(*args, **kwargs):
+            if settings.AWX_DIRECT_SHARED_RESOURCE_MANAGEMENT_ENABLED:
+                return cls.original_delete(*args, **kwargs)
+            raise PermissionDenied({'detail': _('Deletion of this resource is not allowed. Delete this resource via the platform ingress.')})
+
+        cls.delete = delete_wrapper
+
+    if hasattr(cls, 'perform_update'):
+        cls.original_perform_update = cls.perform_update
+
+        @functools.wraps(cls.perform_update)
+        def update_wrapper(*args, **kwargs):
+            if not settings.AWX_DIRECT_SHARED_RESOURCE_MANAGEMENT_ENABLED:
+                view, serializer = args
+                instance = view.get_object()
```

Comment:
Fairly strong preference to replace this line with `instance = serializer.instance`

https://github.com/encode/django-rest-framework/blob/36d5c0e74f562cbe3055f0d20818bd48d3c32359/rest_framework/mixins.py#L66-L68

If you read the origination of `serializer` here, you can see `instance` is already passed into it. The `get_object` method is very non-trivial and expensive and we should avoid re-calling.

Reply:
okay I tried this and it gets tricky
`instance = serializer.instance`
this returns an instance with the incoming data, not the original data. So I can't compare attrs[field] and instance.field

I can access `_prior_values_store`, but those dictionary keys don't match one to one with the `shared_fields`

for example, for a `Team`
```python
 '_prior_values_store': {'description': 'sbf',
                         'name': 'teamA',
                         'organization_id': 1},
```
but the shared_field is called `organization`

I could hard-code `organization` to `organization_id`..but then this code path becomes less generalized.

do you have a suggestion for this? I do agree that re-running a get_object() is expensive

Reply:
> `instance = serializer.instance` this returns an instance populated with new data

In what scenario does this happen? Digging back into DRF, I believe `update` is called for both PATCH and PUT, with the `partial` being True/False depending on which. That `partial` kwarg gets passed to the serializer, but in either case the serializer itself is passed `instance` for the existing instance.

Reply:
sadly we are doing this..

https://github.com/ansible/awx/blob/389a729b759100862ff0450ac0fb0f6a45979651/awx/api/serializers.py#L645

Reply:
You're right, that's grosser than just calling `.get_object`, it's just hamstrung by our legacy serializer. Normal DRF doesn't have this problem of getting a diff, and it's just AWX that's unorthodox.